proxmox/api/ticket_auth.go

```diff
@@ -29,15 +31,91 @@ type ticketAuthenticator struct {
 	mu sync.Mutex
 }
 
+// custom bool to handle bools of: int, text, bool // ref: https://stackoverflow.com/questions/4278430/convert-string-to-integer-type-in-go
+func (bit *cBool) UnmarshalJSON(inBits []byte) error {
+	inString := strings.ToLower(string(inBits))
+
+	inInt, err := strconv.Atoi(inString)
+	if err != nil {
+		return fmt.Errorf("failed injesting bool: %s", inString)
+	}
+
+	switch {
+	case inInt > 0 || inString == "true":
+		*bit = true
+	case inInt == 0 || inString == "false":
+		*bit = false
+	default:
+		return fmt.Errorf("cannot unmarshal boolean: %s", inString)
+	}
+
+	return nil
+}
+
 // NewTicketAuthenticator returns a new ticket authenticator.
+// precedence:  AuthPayload > AuthTicket & CSRFPreventionToken > OTP > user+pass // aka:  pre-auth > OTP > user+pass
 func NewTicketAuthenticator(conn *Connection, creds *Credentials) (Authenticator, error) {
+	if creds.AuthPayload != "" {
+		resBody := AuthenticationResponseBody{}
+
+		err := json.Unmarshal([]byte(creds.AuthPayload), &resBody)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode auth_payload json, %w", err)
+		}
+
+		if resBody.Data.NeedTFA != nil && *(resBody.Data.NeedTFA) {
+			return nil, fmt.Errorf("external auth_payload contains key NeedTFA and requires two-factor authentication")
+		}
+
+		creds.Username = resBody.Data.Username                              // todo: exclude line? - is creds.Username needed/used anywhere other than new-auth reqs?
+		if creds.Username != "" && !strings.Contains(creds.Username, "@") { // todo: improve this vs copy-pasta from credentials.go
```

Comment:
dont think thes 2x `creds.Username = ..` are needed (here and L91) - but want have higher confidence before removing them

Reply:
updated inline comment; added git-comment to new-pointer